lib/elixir/lib/kernel.ex

```diff
@@ -2173,6 +2173,15 @@ defmodule Kernel do
     Enum.reduce(fields, struct, fun)
   end
 
+  defp validate_struct!(%{__struct__: module} = struct, _module, _arity) when is_atom(module) do
+    struct
+  end
+
+  defp validate_struct!(expr, module, arity) do
+    error_message = :elixir_map.format_error({:invalid_struct_return_value, module, arity, expr})
```

Comment:
This looks like too much work! 😅 I would just duplicate the error message. Thoughts?

Reply:
I also had doubts.

If I duplicate the error message I would need to `:elixir_aliases.inspect(module)` and `Macro.to_string(expr)`, which more or less balances with the `:elixir_map.format_error` and `:erlang.iolist_to_binary`. And this way the error message are in sync.

But I don't have a strong opinion here. Your call 🙂

Reply:
On this side of things, it is ok to use inspect + Macro.to_string. We can keep this more Elixir looking than the rest.